lib/widgets/set_status.dart

```diff
@@ -0,0 +1,339 @@
+          //      text field.
+          if (text == statusTextController.text) return;
+
+          statusTextController.text = text;
+        });
+  }
+
+  @override
+  void dispose() {
+    statusTextController.dispose();
+    statusChange.dispose();
+    super.dispose();
+  }
+
+  List<UserStatus> statusSuggestions(BuildContext context) {
+    final store = PerAccountStoreWidget.of(context);
+    final localizations = ZulipLocalizations.of(context);
+
+    final values = [
+      (localizations.userStatusBusy,            '1f6e0'),
+      (localizations.userStatusInAMeeting,      '1f4c5'),
+      (localizations.userStatusCommuting,       '1f68c'),
+      (localizations.userStatusOutSick,         '1f912'),
+      (localizations.userStatusVacationing,     '1f334'),
+      (localizations.userStatusWorkingRemotely, '1f3e0'),
+      (localizations.userStatusAtTheOffice,     '1f3e2'),
+    ];
+
+    return [
+      for (final (statusText, emojiCode) in values)
+        if (store.getUnicodeEmojiNameByCode(emojiCode) case final emojiName?)
+          UserStatus(
+            text: statusText,
+            emoji: StatusEmoji(
+              emojiName: emojiName,
+              emojiCode: emojiCode,
+              reactionType: ReactionType.unicodeEmoji)),
+    ];
+  }
+
+  void handleStatusClear() {
+    statusChange.value = UserStatusChange(
+      text: toOption(recent: null, old: oldStatus.text),
+      emoji: toOption(recent: null, old: oldStatus.emoji),
+    );
+  }
+
+  Future<void> handleStatusSave() async {
+    final store = PerAccountStoreWidget.of(context);
+    final localizations = ZulipLocalizations.of(context);
+
+    Navigator.pop(context);
+    if (newStatus == oldStatus) return;
+
+    try {
+      await updateStatus(store.connection, change: statusChange.value);
+    } catch (e) {
+      reportErrorToUserBriefly(localizations.updateStatusErrorTitle);
+    }
+  }
+
+  void chooseStatusEmoji() async {
+    final emojiCandidate = await showEmojiPickerSheet(pageContext: context);
+    if (emojiCandidate == null) return;
+
+    final emoji = StatusEmoji(
+      emojiName: emojiCandidate.emojiName,
+      emojiCode: emojiCandidate.emojiCode,
+      reactionType: emojiCandidate.emojiType,
+    );
+    statusChange.value = statusChange.value.copyWith(
+      emoji: toOption(recent: emoji, old: oldStatus.emoji)
+    );
+  }
+
+  void chooseStatusSuggestion(UserStatus status) {
+    statusChange.value = UserStatusChange(
+      text: toOption(recent: status.text, old: oldStatus.text),
+      emoji: toOption(recent: status.emoji, old: oldStatus.emoji)
+    );
+  }
+
+  Option<T> toOption<T>({required T recent, required T old}) =>
+    recent == old ? OptionNone() : OptionSome(recent);
+
+  @override
+  Widget build(BuildContext context) {
+    final designVariables = DesignVariables.of(context);
+    final localizations = ZulipLocalizations.of(context);
+    final suggestions = statusSuggestions(context);
+
+    return Scaffold(
```

Comment:
nit: this is a different kind of thing from the steps above (they're quick O(1) data lookups, while this is doing a bit of computation of its own), so set it off with a blank line:
```suggestion
    final designVariables = DesignVariables.of(context);
    final localizations = ZulipLocalizations.of(context);

    final suggestions = statusSuggestions(context);

    return Scaffold(
```